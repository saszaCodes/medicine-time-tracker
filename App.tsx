import { ThemeProvider } from "styled-components/native";
import { SettingsProvider } from "./contexts/SettingsContext";
import { TimeTrackerProvider } from "./contexts/TimeTrackerContext";
import { MainNavigator } from "./navigators/MainNavigator";
import { theme } from "./styles/theme";

export default function App() {
  return (
    <SettingsProvider>
      <TimeTrackerProvider>
        <ThemeProvider theme={theme}>
          <MainNavigator />
        </ThemeProvider>
      </TimeTrackerProvider>
    </SettingsProvider>
  );
}
