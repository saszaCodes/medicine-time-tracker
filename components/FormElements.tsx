import {
  Picker,
  PickerProps,
  PickerItemProps,
} from "@react-native-picker/picker";
import { PropsWithChildren } from "react";
import { ButtonProps, PressableProps, TextInputProps } from "react-native";
import styled from "styled-components/native";

const FormItemContainer = styled.View`
  margin-top: ${({ theme }) => theme.spacing.xxxl};
`;
const FormItemlabelContainer = styled.View`
  margin-bottom: ${({ theme }) => theme.spacing.mplus};
`;
const FormItemLabel = styled.Text`
  font: ${({ theme }) => theme.fonts.subheading2};
`;
export const FormItemWrapper = (
  props: PropsWithChildren<{ label?: string }>
) => (
  <FormItemContainer>
    {props.label && (
      <FormItemlabelContainer>
        <FormItemLabel>{props.label}</FormItemLabel>
      </FormItemlabelContainer>
    )}
    {props.children}
  </FormItemContainer>
);

const Input = styled.TextInput`
  font: ${({ theme }) => theme.fonts.regular};
  background-color: ${({ theme }) => theme.colors.backgroundSecondary};
  padding: ${({ theme }) => theme.spacing.s} ${({ theme }) => theme.spacing.l};
  border-radius: ${({ theme }) => theme.borderRadi.regular};
  // TODO: add inset shadow
`;
export const TextInput = ({
  label,
  ...props
}: TextInputProps & { label?: string }) => (
  <FormItemWrapper {...{ label }}>
    <Input {...props} />
  </FormItemWrapper>
);

type PicklistPropsType<T> = PickerProps & {
  options: (Omit<PickerItemProps, "value"> & { value: T })[];
};
// type PicklistPropsType = PickerProps;
export const Picklist: <T = any>(props: PicklistPropsType<T>) => JSX.Element = (
  props
) => (
  <Picker {...props}>
    {props.options.map((itemProps, i) => (
      <Picker.Item key={i} {...itemProps} />
    ))}
  </Picker>
);

const ButtonContainer = styled.View`
  flex-direction: row;
  margin-bottom: ${({ theme }) => theme.spacing.mplus};
  justify-content: center;
  // TODO: make button not 100% wide
`;

const StyledButton = styled.Pressable`
  font: ${({ theme }) => theme.fonts.regular};
  color: ${({ theme }) => theme.colors.text};
  background-color: ${({ theme }) => theme.colors.primary};
  border-radius: ${({ theme }) => theme.borderRadi.regular};
  padding: ${({ theme }) => theme.spacing.m} ${({ theme }) => theme.spacing.l};
`;

const ButtonText = styled.Text``;

export const Button = (
  props: Pick<PressableProps, "onPress"> & { title: string }
) => (
  <FormItemWrapper>
    <ButtonContainer>
      <StyledButton onPress={props.onPress}>
        <ButtonText>{props.title}</ButtonText>
      </StyledButton>
    </ButtonContainer>
  </FormItemWrapper>
);
