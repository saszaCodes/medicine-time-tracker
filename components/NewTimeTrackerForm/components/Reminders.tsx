import Checkbox from "expo-checkbox";
import { Text, View } from "react-native";
import { parseTimePeriods } from "../../../utils/parseTimePeriods";

type RemindersProps = {
  initialValue?: ReminderKeys;
  reminders: ReminderKeys;
  onChange: (reminders: ReminderKeys) => void;
};
export type ReminderOptions = typeof reminderOptions;
export type ReminderKey = keyof ReminderOptions;
export type ReminderKeys = ReminderKey[];

export const reminderOptions = {
  "15min": {
    label: "15 minutes before",
    ms: parseTimePeriods(15, "minute(s)", "ms"),
  },
  "30min": {
    label: "30 minutes before",
    ms: parseTimePeriods(30, "minute(s)", "ms"),
  },
  "1h": {
    label: "1 hour before",
    ms: parseTimePeriods(1, "hour(s)", "ms"),
  },
  "3h": {
    label: "3 hours before",
    ms: parseTimePeriods(3, "hour(s)", "ms"),
  },
  "1d": {
    label: "1 day before",
    ms: parseTimePeriods(1, "day(s)", "ms"),
  },
  "3d": {
    label: "3 days before",
    ms: parseTimePeriods(3, "day(s)", "ms"),
  },
} as const;

const reminderKeys = Object.keys(reminderOptions) as ReminderKeys;

export const Reminders = (props: RemindersProps) => {
  const handleValueChange = (key: ReminderKey, newValue: boolean) => {
    if (newValue) props.onChange([...props.reminders, key]);
    else
      props.onChange(
        props.reminders.filter((reminderKey) => reminderKey !== key)
      );
  };

  const generateReminders = () =>
    reminderKeys.map((key, index) => {
      const { label } = reminderOptions[key];
      return (
        <View key={index}>
          <Checkbox
            value={props.reminders.includes(key)}
            onValueChange={(value) => handleValueChange(key, value)}
          />
          <Text>{label}</Text>
        </View>
      );
    });

  return <>{generateReminders()}</>;
};
