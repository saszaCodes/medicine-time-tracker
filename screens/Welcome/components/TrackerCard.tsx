import moment from "moment";
import { View } from "react-native";
import styled from "styled-components/native";
import { Tracker } from "../../../types/types";
import { PropsWithChildren } from "react";

const Container = styled.View`
  margin-top: ${({ theme }) => theme.spacing.mplus};
  // TODO: get ":last-of-type" working
  :last-of-type {
    margin-bottom: ${({ theme }) => theme.spacing.mplus};
    border: 1px solid red;
  }
  padding: ${({ theme }) => theme.spacing.m};
  border-radius: ${({ theme }) => theme.borderRadi.regular};
  flex-direction: row;
  align-items: center;
  min-height: 80px;
  background-color: ${({ theme }) => theme.colors.backgroundSecondary};
  // TODO: get shadow working
  box-shadow: 5px 4px ${({ theme }) => theme.colors.boxShadow};
`;

const DateContainer = styled.View`
  margin-right: ${({ theme }) => theme.spacing.l};
`;

const Heading = styled.Text`
  font: ${({ theme }) => theme.fonts.subheading1};
  font-weight: 800;
`;

type TrackerCardProps = { icon?: JSX.Element; tracker: Tracker };

export const TrackerCard = (props: TrackerCardProps) => {
  const {
    icon,
    tracker: { name, finishDate, description, reminders },
  } = props;
  return (
    <Container>
      <DateContainer>
        <Heading>{moment(finishDate).format(" hh:mm")}</Heading>
      </DateContainer>
      <View>
        <Heading>{name}</Heading>
      </View>
    </Container>
  );
};
