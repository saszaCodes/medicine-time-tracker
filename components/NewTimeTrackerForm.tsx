import { useNavigation } from "@react-navigation/native";
import { useEffect } from "react";
import { useForm, Controller, FieldValues } from "react-hook-form";
import { Button } from "react-native";
import {
  initialTrackerFormData,
  useTimeTrackerContext,
} from "../contexts/TimeTrackerContext";
import {
  FinishTimeType,
  TimePeriod,
  Tracker,
  TrackerFormInput,
} from "../types/types";
import { parseTimePeriods } from "../utils/parseTimePeriods";
import { Picklist, TextInput } from "./FormElements";

type NewTimeTrackerFormProps = {
  useContextValues?: boolean;
  display?: {
    [key in keyof TrackerFormInput]?: boolean | string;
  };
};

// TODO: get rid of any time hidden in FieldValues
// Overload signatures for parseValues
function parseValues(values: FieldValues, parseTo?: "Tracker"): Tracker;
function parseValues(
  values: FieldValues,
  parseTo: "TrackerFormInput"
): TrackerFormInput;
// Parse values to either Tracker or TrackerFormInput (by default: Tracker)
function parseValues(
  values: FieldValues,
  parseTo: "TrackerFormInput" | "Tracker" = "Tracker"
) {
  // Define props common for both return types
  const parsedValues = {
    name: values.name,
    description: values.description,
    reminders: values.reminders,
  };
  // Add a prop specific to Tracker type and return
  if (parseTo === "Tracker") {
    const finishDate =
      values.finishTimeType === "date"
        ? values.finishTimeValue
        : parseTimePeriods(
            values.finishTimeValue,
            values.finishTimePeriod,
            "ms"
          ) + Date.now();
    console.log(`finishDate: ${finishDate}`);
    return { ...parsedValues, finishDate } as Tracker;
  }
  // Add a prop specific to TrackerFormInput type and return
  else {
    const finishTime: TrackerFormInput["finishTime"] = {
      type: values.finishTimeType,
      period: values.finishTimePeriod,
      value: values.finishTimeValue,
    };
    console.log(`finishTime: ${finishTime}`);
    return { ...parsedValues, finishTime } as TrackerFormInput;
  }
}

// TODO: don't use context values if useContextValues = false
export const NewTimeTrackerForm = ({
  useContextValues = true,
  display = {
    description: true,
    finishTime: true,
    name: true,
    reminders: true,
  },
}: NewTimeTrackerFormProps) => {
  const { control, handleSubmit, getValues, watch, setValue } = useForm();
  const {
    addTracker,
    trackerFormData,
    updateTrackerFormData,
    resetTrackerFormData,
  } = useTimeTrackerContext();
  const navigation = useNavigation();

  // When form goes out of focus, update the context (unless useContextValues is false)
  useEffect(() => {
    // TODO: analyze how this useEffect affects when component is rerendered and what side effects it causes
    // (hint: add console.log() in component body)
    const subscription = navigation.addListener("blur", () => {
      console.log("Form context updated");
      if (!useContextValues) return;
      const values = getValues();
      const newFormData = parseValues(values, "TrackerFormInput");
      updateTrackerFormData(newFormData);
    });

    return navigation.removeListener("blur", subscription);
  }, []);

  const finishTimeSelectType: FinishTimeType = watch("finishTimeType");

  useEffect(() => {
    console.log(finishTimeSelectType);
  }, [finishTimeSelectType]);

  // On first render read initial form data from context
  useEffect(() => {
    // Extract form data from context (unless useContextValues is false)
    const {
      name,
      description,
      reminders,
      finishTime: {
        type: finishTimeType,
        period: finishTimePeriod,
        value: finishTimeValue,
      },
    } = useContextValues ? trackerFormData : initialTrackerFormData;
    // Set initial form values
    setValue("name", name);
    setValue("description", description);
    setValue("reminders", reminders?.toString());
    setValue("finishTimeType", finishTimeType);
    setValue("finishTimePeriod", finishTimePeriod);
    setValue("finishTimeValue", finishTimeValue);
  }, []);

  // On submit clear form context (unless useContextValues is false) and add a new tracker
  const onSubmit = () => {
    console.log("Form submitted");
    const values = getValues();
    const newTracker = parseValues(values);
    addTracker(newTracker);
    if (!useContextValues) return;
    resetTrackerFormData();
  };

  //  Return a form consisting of controlled elements
  return (
    <>
      {display.name && (
        <Controller
          name="name"
          control={control}
          rules={{ required: true }}
          render={({ field: { value, onChange } }) => (
            <TextInput
              value={value}
              onChangeText={onChange}
              placeholder="Name"
            />
          )}
        />
      )}
      {display.description && (
        <Controller
          name="description"
          control={control}
          render={({ field: { value, onChange } }) => (
            <TextInput
              multiline
              value={value}
              onChangeText={onChange}
              placeholder="Description"
            />
          )}
        />
      )}
      {display.reminders && (
        <Controller
          rules={{ pattern: /[0-9]+/ }}
          name="reminders"
          control={control}
          render={({ field: { value, onChange } }) => (
            <TextInput
              value={value}
              onChangeText={onChange}
              placeholder="Reminders"
            />
          )}
        />
      )}
      {display.finishTime && (
        // TODO: add more options for particular timer periods
        <>
          <Controller
            name="finishTimeType"
            control={control}
            rules={{ required: true }}
            render={({ field: { onChange, value } }) => (
              <Picklist<FinishTimeType>
                selectedValue={value}
                onValueChange={onChange}
                options={[
                  { label: "Date", value: "date" },
                  { label: "TimePeriod", value: "timePeriod" },
                ]}
              />
            )}
          />
          {finishTimeSelectType === "timePeriod" && (
            <Controller
              name="finishTimePeriod"
              rules={{ required: true }}
              control={control}
              render={({ field: { onChange, value } }) => (
                <Picklist<TimePeriod>
                  selectedValue={value}
                  onValueChange={onChange}
                  options={[
                    { label: "Minute(s)", value: "minute(s)" },
                    { label: "Hour(s)", value: "hour(s)" },
                    { label: "Day(s)", value: "day(s)" },
                    { label: "Week(s)", value: "week(s)" },
                  ]}
                />
              )}
            />
          )}
          <Controller
            rules={{ pattern: /[0-9]+/, required: true }}
            name="finishTimeValue"
            control={control}
            render={({ field: { value, onChange } }) => (
              <TextInput
                value={value}
                onChangeText={onChange}
                placeholder="Finish time"
              />
            )}
          />
        </>
      )}
      <Button onPress={handleSubmit(onSubmit)} title="Save" />
    </>
  );
};
