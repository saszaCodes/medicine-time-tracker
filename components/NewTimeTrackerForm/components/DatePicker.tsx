import { Text, View } from "react-native";
import {
  AndroidNativeProps,
  DateTimePickerAndroid,
} from "@react-native-community/datetimepicker";
import styled from "styled-components/native";
import { FormItemWrapper } from "../../FormElements";

export {};

type DatePickerProps = {
  value: number;
  onChange: (selectedDate: number) => void;
};

const extractDateInfo = (date: Date) => ({
  year: date.getFullYear(),
  month: date.getMonth(),
  day: date.getDate(),
  hour: date.getHours(),
  minutes: date.getMinutes(),
});

export const DatePicker = (props: DatePickerProps) => {
  const date = new Date(props.value);
  const { year, month, day, hour, minutes } = extractDateInfo(date);

  const openPicker = (
    params: Omit<AndroidNativeProps, "onChange" | "value">
  ) => {
    DateTimePickerAndroid.open({
      ...params,
      value: date,
      onChange: (event, selectedDate) => {
        DateTimePickerAndroid.dismiss(params.mode);
        if (selectedDate === undefined || event.type === "dismissed") return;
        props.onChange(selectedDate.valueOf());
      },
    });
  };

  const Container = styled.View`
    flex-direction: row;
  `;

  const Text = styled.Text`
    font: ${({ theme }) => theme.fonts.regular};
    background-color: ${({ theme }) => theme.colors.backgroundSecondary};
    padding: ${({ theme }) => theme.spacing.s} ${({ theme }) => theme.spacing.l};
    border-radius: ${({ theme }) => theme.borderRadi.regular};
    // TODO: remove right margin from last item
    margin-right: ${({ theme }) => theme.spacing.mplus};
    // TODO: add inset shadow
  `;

  return (
    <FormItemWrapper label="Finish date">
      <Container>
        <Text
          onPress={() => openPicker({ mode: "date" })}
        >{`${day}/${month}/${year}`}</Text>
        <Text onPress={() => openPicker({ mode: "time" })}>{`${hour}:${
          minutes < 10 ? 0 : ""
        }${minutes}`}</Text>
      </Container>
    </FormItemWrapper>
  );
};
