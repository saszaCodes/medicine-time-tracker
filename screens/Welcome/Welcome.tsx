import { DrawerScreenProps } from "@react-navigation/drawer";
import { useCallback } from "react";
import { Button, View, Text } from "react-native";
import { ScrollView } from "react-native-gesture-handler";
import { reminderOptions } from "../../components/NewTimeTrackerForm/components/Reminders";
import { useTimeTrackerContext } from "../../contexts/TimeTrackerContext";
import { MainNavigatorParams } from "../../types/types";
import moment from "moment";

// TODO: fix type
export const WelcomeScreen = ({
  navigation,
}: DrawerScreenProps<MainNavigatorParams>) => {
  const { trackers } = useTimeTrackerContext();

  const generateTrackers = useCallback(
    (params?: { startDate: number; endDate: number }) => {
      return trackers.map((tracker, i) => {
        const { name, finishDate, description, reminders } = tracker;
        if (params) {
          const { startDate: listStartDate, endDate: listEndDate } = params;
          if (listStartDate >= listEndDate) return null;
          if (listStartDate > finishDate || finishDate > listEndDate)
            return null;
        }
        return (
          <View key={i}>
            <Text>{name}</Text>
            <Text>{finishDate}</Text>
            <Text>{description}</Text>
            {reminders
              ? reminders?.map((key, i) => (
                  <Text key={i}>{reminderOptions[key].label}</Text>
                ))
              : null}
          </View>
        );
      });
    },
    [trackers]
  );

  return (
    <ScrollView>
      <Text>Today</Text>
      {generateTrackers({
        startDate: moment().valueOf(),
        endDate: moment().endOf("day").valueOf(),
      })}
      <Text>Tommorow</Text>
      {generateTrackers({
        startDate: moment().add(1, "days").startOf("day").valueOf(),
        endDate: moment().add(1, "days").endOf("day").valueOf(),
      })}
      <Text>Missed</Text>
      {generateTrackers({
        startDate: 0,
        endDate: moment().valueOf(),
      })}
      <Text>Future</Text>
      {generateTrackers({
        startDate: moment().add(1, "days").endOf("day").valueOf(),
        endDate: Infinity,
      })}
      <Button
        onPress={() => navigation.navigate("AddTracker")}
        title="Add new tracker"
      />
    </ScrollView>
  );
};
