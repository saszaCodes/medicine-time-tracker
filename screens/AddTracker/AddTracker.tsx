import { NewTimeTrackerForm } from "../../components/NewTimeTrackerForm/NewTimeTrackerForm";

export const AddTrackerScreen = () => {
  return <NewTimeTrackerForm />;
};
