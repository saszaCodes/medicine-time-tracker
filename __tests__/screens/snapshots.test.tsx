import { Text } from "react-native";
import renderer from "react-test-renderer";
import App, { Contexts } from "../../App";
// SCREENS
// Welcome screen and related components
import { WelcomeScreen } from "../../screens/Welcome/Welcome";
import { TrackerCard } from "../../screens/Welcome/components/TrackerCard";
// Settings screen
import { SettingsScreen } from "../../screens/Settings/Settings";
// AddTracker screen
import { AddTrackerScreen } from "../../screens/AddTracker/AddTracker";

describe("Snapshot matches: ", () => {
  it("WelcomeScreen", () => {
    const component = renderer
      .create(
        <Contexts>
          <WelcomeScreen />
        </Contexts>
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });
});
