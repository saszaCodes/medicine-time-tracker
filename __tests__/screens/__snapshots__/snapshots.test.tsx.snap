// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot matches:  WelcomeScreen 1`] = `
<Text>
  My app
</Text>
`;
